app/utils/kci_test/regressions.py

```diff
@@ -0,0 +1,466 @@
+                                             test_group_doc,
+                                             db_options)
+        if prev_reg_doc[0]:
+            doc_id = prev_reg_doc[0]
+
+            regr_data_key = \
+                REGRESSION_DOT_FMT.format(models.REGRESSIONS_KEY, regr_key)
+
+            if prev_reg_doc[1]:
+                # If we also have the same key in the document, append the
+                # new test group report.
+                document = {"$addToSet": {regr_data_key: test_group_doc}}
+            else:
+                # Otherwise just set the new key.
+                document = {"$set": {regr_data_key: regr_docs}}
+
+            ret_val = utils.db.update3(
+                models.TEST_GROUP_REGRESSIONS_COLLECTION,
+                {models.ID_KEY: prev_reg_doc[0]},
+                document,
+                db_options=db_options
+            )
+        else:
+            regression_doc = {
+                models.CREATED_KEY: created_on,
+                models.GIT_BRANCH_KEY: branch,
+                models.JOB_ID_KEY: job_id,
+                models.JOB_KEY: job,
+                models.KERNEL_KEY: kernel,
+                models.NAME_KEY: name
+            }
+
+            # The regression data structure.
+            # A dictionary with nested keys, whose keys in nested order are:
+            # lab name
+            # architecture type
+            # board name
+            # board instance or the string "none"
+            # defconfig full string
+            # compiler string (just compiler + version)
+            # The regressions are stored in a list as the value of the
+            # "compiler" key.
+            regression_doc[models.REGRESSIONS_KEY] = {
+                lab: {
+                    arch: {
+                        board: {
+                            b_instance: {
+                                defconfig: {
+                                    compiler: regr_docs
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            ret_val, doc_id = \
+                utils.db.save3(
+                    models.TEST_GROUP_REGRESSIONS_COLLECTION, regression_doc,
+                    db_options=db_options)
+
+        # Save the regressions id and test group id in an index collection.
+        if all([any([ret_val == 201, ret_val == 200]), doc_id]):
+            utils.db.save3(
+                models.TEST_GROUP_REGRESSIONS_BY_TEST_GROUP_COLLECTION,
+                {
+                    models.TEST_GROUP_ID_KEY: test_group_id,
+                    models.TEST_GROUP_REGRESSIONS_ID_KEY: doc_id,
+                    models.CREATED_KEY: created_on
+                },
+                db_options=db_options
+            )
+
+    return ret_val, doc_id
+
+
+def check_and_track(test_group_doc, db_options):
+    """Check previous test group report and start tracking regressions.
```

Comment:
I think it's worth mentioning that this function doesn't recursively check for regressions in sub-groups, so the caller has to do the job.

Reply:
It looks like this `check_and_track` function only tracks regressions for the test cases inside the given `test_group_doc`, but it doesn't look for sub-groups inside it.  So to track all regressions, this function must be called again for each sub-group and their own sub-groups recursively.  An alternative implementation would be to do the recursion in `check_and_track` so it would only need to be called once.